src/runtime/parcelset/detail/parcel_await.cpp

```diff
@@ -15,31 +15,39 @@
 namespace hpx { namespace parcelset { namespace detail {
 
     parcel_await::parcel_await(parcel&& p, int archive_flags,
-        parcel_await::put_parcel_type pp)
+            parcel_await::put_parcel_type pp)
       : put_parcel_(std::move(pp)),
-        archive_(preprocess_, archive_flags),
-        overhead_(archive_.bytes_written()),
+        archive_flags_(archive_flags),
         idx_(0)
     {
         parcels_.push_back(std::move(p));
+        handled_.push_back(0);
     }
 
     parcel_await::parcel_await(std::vector<parcel>&& parcels, int archive_flags,
-        parcel_await::put_parcel_type pp)
+            parcel_await::put_parcel_type pp)
       : put_parcel_(std::move(pp)),
         parcels_(std::move(parcels)),
-        archive_(preprocess_, archive_flags),
-        overhead_(archive_.bytes_written()),
+        handled_(parcels_.size(), 0),
+        archive_flags_(archive_flags),
         idx_(0)
-    {
-    }
+    {}
 
     void parcel_await::apply()
     {
         for (/*idx_*/; idx_ != parcels_.size(); ++idx_)
         {
-            archive_.reset();
-            archive_ << parcels_[idx_];
+            // send every parcel only once
+            if (handled_[idx_])
```

Comment:
I think that's wrong. parcels need to get preprocessed as many times as there are no futures to be waited on. For example, a shared_future<hpx::id_type> inside a parcels might need to be proprocessed three times. First, the future is not ready and needs to be waited on, second (now the future is ready) the id_type needs to be split (another future needs to be waited on), the third one is the last in the fixed point iteration to ensure everything is ready and the last round did indeed create a value and not an exception, that is, to calculate the final size of the data to be sent.